internal/controller/ws_controller_test.go

```diff
@@ -0,0 +1,96 @@
+package controller
+
+import (
+	"encoding/base64"
+	"net/http/httptest"
+	"strings"
+	"testing"
+
+	"github.com/RedHatInsights/platform-receptor-controller/internal/receptor/protocol"
+	"github.com/gorilla/mux"
+	"github.com/gorilla/websocket"
+	"github.com/stretchr/testify/assert"
+	"go.uber.org/goleak"
+)
+
+const validJSON = `{ "identity": {"account_number": "540155", "type": "User", "internal": { "org_id": "1979710" } } }`
+
+func newServer(t *testing.T, useRouting bool) *httptest.Server {
+	rc := NewReceptorController(NewConnectionManager(), mux.NewRouter())
+	rc.Routes()
+
+	if useRouting == false {
+		s := httptest.NewServer(rc.handleWebSocket())
+		s.URL = "ws" + strings.TrimPrefix(s.URL, "http")
+		return s
+	}
+
+	s := httptest.NewServer(rc.router)
+	s.URL = "ws" + strings.TrimPrefix(s.URL, "http")
+	return s
+}
+
+// This test is just to verify that we can upgrade a connection and not have any leaking goroutines after it closes.
+func TestHandshake(t *testing.T) {
+	defer goleak.VerifyNone(t)
+
+	s := newServer(t, false)
+	defer s.Close()
+
+	ws, _, err := websocket.DefaultDialer.Dial(s.URL, nil)
+	if err != nil {
+		t.Fatalf("%v", err)
+	}
+	defer ws.Close()
+
+	hiMessage := protocol.HiMessage{Command: "HI", ID: "test-node"}
+	w, err := ws.NextWriter(websocket.BinaryMessage)
+	if err := protocol.WriteMessage(w, &hiMessage); err != nil {
+		t.Fatalf("Test Write failed to write HI: %v", err)
+	}
+	w.Close()
+
+	messageType, r, err := ws.NextReader()
+	if err != nil {
+		t.Fatalf("Test Reader failed to read response: %v", err)
+	}
+	assert.Equal(t, messageType, websocket.BinaryMessage, "The Hi response from the server should be a binary message")
+
+	message, err := protocol.ReadMessage(r)
+	if err != nil {
+		t.Fatalf("Protocol Reader failed to read the response: %v", err)
+	}
+	assert.Equal(t, message.Type(), protocol.HiMessageType, "The Hi response should be of type protocol.HiMessageType")
+}
+
+// This test is verifying that the router is working correctly.
+func TestRouting(t *testing.T) {
```

Comment:
I would split this test up into multiple tests.  I think splitting the test up would clarify things a bit and make it easier to track down test failures.  This is a bit nit picky as I know we are still experimenting / playing around with testing at this point.  Ignore this if you are currently rewriting this test using the other testing framework, etc.

Can you add a test where the header is invalid json and/or causes the header parsing to fail within the platform-middleware code?

Reply:
Np! Currently rewriting this one.